### PR TITLE
feat: add isGurmukhi function to check if string is Unicode Gurmukhi

### LINKED ISSUE #140: Find Unicode Gurmukhi instead of ASCII in Search query in Lines Model
Currently, `is-ascii` is being used to detect if the search query is in the GurbaniAkhar format. 

However, GurbaniAkhar has letters that are outside the 128char ASCII range. This triggers the search to convert Unicode -> GurbaniAkhar, which messes up the search query (esp in Full Word searches).

Solutions:

1. Find a library that can check if a character is in a unicode block

2. Build own solution in `gurmukhi-utils` to check if string has Unicode Gurmukhi

### PR DESCRIPTION
**Pull request by Sarabveer**
Closes #140

### CODE REVIEW

**Comment by Sarabveer** [comment]
Need to add tests

**Comment by Harjot1Singh** [comment]
I'd suggest calling it `isUnicodeGurmukhi` or `isGurmukhiUnicode`, since `isGurmukhi` can be easily confused with "Is this string Gurmukhi" btw.

File: lib/isGurmukhi.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Checks if first char in string is part of the Gurmukhi Unicode block.
+ * @param {String} text The text to check.
+ * @return {boolean} True if Unicode Gurmukhi, false if other.
+ * @example
+ * isGurmukhi('ਗੁਰਮੁਖੀ') // => true
+ * isGurmukhi('gurmuKI') // => false
+ */
+const isGurmukhi = text => ( text.charCodeAt( 0 ) >= 2560 && text.charCodeAt( 0 ) <= 2687 )
```
Comment (Harjot1Singh) [inline lib/isGurmukhi.js:9]: Is testing the first char sufficient? Seems a little cheeky lol. I guess it's sufficient for most cases, but maybe add a second parameter, `exhaustive [bool]` which will check every character in the string if `true`.

I do think that checking every character might be fine though, perf-wise, unless there's some mega long strings being passed in.

**Comment by Sarabveer** [comment]
> 
> 
> I'd suggest calling it `isUnicodeGurmukhi` or `isGurmukhiUnicode`, since `isGurmukhi` can be easily confused with "Is this string Gurmukhi" btw.

ASCII is ASCII, it isn't Gurmukhi. There isn't a way to test if a string is from an ASCII Gurmukhi font, and if there was, English would return true for such a function.

File: lib/isGurmukhi.js (hunk 1 above)
Comment (Sarabveer) [reply lib/isGurmukhi.js:9]: That can be done.

File: index.d.ts [hunk 2]
```diff
@@ -8,4 +8,6 @@ export function toHindi(text: string): string
 
 export function toShamukhi(text: string): string
 
-export function firstLetters(text: string, stripNukta?: boolean = true, withVishraams?: boolean): string
\ No newline at end of file
+export function firstLetters(text: string, stripNukta?: boolean = true, withVishraams?: boolean): string
+
+export function isGurmukhi(text: string, exhaustive?: boolean ): boolean
```
Comment (Harjot1Singh) [inline index.d.ts:13]: Fix formatting here, eslint can’t read TS

File: test/isGurmukhi.spec.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  [ 'ਮੈਂ ਗੁਰਮੁਖੀ ਵਿਚ ਲਿਖ ਰਿਹਾ ਹਾਂ।', true ],
+  [ 'ਲੜੀਵਾਰ​ਗੁਰਬਾਣੀ', true ], // Has U+200B, Zero Width Space
+  [ 'मैं हिंदी में लिख रहा हूँ।', false ],
+  [ 'میں شاہ رخ میں لکھ رہا ہوں۔', false ],
+  [ 'ਗੁਰਮੁਖੀ & English', false ],
+]
+
+describe( 'isGurmukhi()', () => {
+  strings.map( ( [ string, result ] ) => it( `should convert '${string}' to ${result}`, () => {
+    expect( isGurmukhi( string, true ) ).to.equal( result )
```
Comment (Harjot1Singh) [inline test/isGurmukhi.spec.js:17]: Add test case for when the second Param isn’t supplied. It’ll bump us back to 100% coverage.

File: docs/README.md [hunk 4]
```diff
@@ -92,6 +93,20 @@ firstLetters('iZir&qym sMdUk* drIXw AmIk* ]', false) // => Zsda
...
+
+| Param | Type | Description |
+| --- | --- | --- |
+| text | <code>String</code> | The text to check. |
+| [exhaustive] | <code>boolean</code> | If `true`, checks if the whole string is Unicode Gurmukhi. |
+
+**Example**  
+```js
+isGurmukhi('ਗੁਰਮੁਖੀ') // => true
isGurmukhi('gurmuKI') // => false
```
Comment (Harjot1Singh) [inline docs/README.md:108]: Formatting output seems
To be wrong here. Check the other funcs to see how to format examples. Please also add an example to examples.js too

**Comment by bhajneet** [comment]
Since this PR is closing #140, perhaps we can also fix toAscii function in this PR?

**Comment by Sarabveer** [comment]
> 
> 
> Since this PR is closing #140, perhaps we can also fix toAscii function in this PR?

@Harjot1Singh how would this be done?

**Comment by Harjot1Singh** [comment]
> > Since this PR is closing #140, perhaps we can also fix toAscii function in this PR?
> 
> @Harjot1Singh how would this be done?

`toAscii = (text) => isAscii(text) ? text : otherStuffInFunction`

Somrething like that lol

**Comment by bhajneet** [comment]
> 
> 
> > > Since this PR is closing #140, perhaps we can also fix toAscii function in this PR?
> > 
> > 
> > @Harjot1Singh how would this be done?
> 
> `toAscii = (text) => isAscii(text) ? text : otherStuffInFunction`
> 
> Something like that. I'm not sure if we'd want exhaustive enabled or not.

Wouldn't it be better to fix `toAscii` from not running the post rules block if it's already in ascii?

**Comment by Harjot1Singh** [comment]
> > > > Since this PR is closing #140, perhaps we can also fix toAscii function in this PR?
> > > 
> > > 
> > > @Harjot1Singh how would this be done?
> > 
> > 
> > `toAscii = (text) => isAscii(text) ? text : otherStuffInFunction`
> > Something like that. I'm not sure if we'd want exhaustive enabled or not.
> 
> Wouldn't it be better to fix `toAscii` from not running the post rules block if it's already in ascii?

Wouldn't this mean running some of `toAscii`, followed by then running `isAscii`? Why do extra work?

**Comment by Sarabveer** [comment]
There is no `isAscii` it is `isGurmukhi`

**Comment by Harjot1Singh** [comment]
> There is no `isAscii` it is `isGurmukhi`

My bad, `isGurmukhi` instead of `isAscii`

**Comment by bhajneet** [comment]
> Wouldn't this mean running some of `toAscii`, followed by then running `isAscii`? Why do extra work?

Sorry, I misread. Using isExtendedAscii/isGurmukhi would be fine in the line you said. However since `isGurmukhi`'s exhaustive parameter requires all characters to be in unicode to pass, I do not believe it will work for shabados/desktop. (In the sense that you are trying to convert something toAscii on each keystroke of unicode keyboard and each keystroke is supposed to be converted toAscii).

**Comment by bhajneet** [comment]
If you use "some" instead of "every" in `isGurmukhi` function then it isn't true to the name of the function. Maybe better to check "every"/exhaustive for extended ascii chars and if false then converting to ascii?

**Comment by Sarabveer** [comment]
> However since `isGurmukhi`'s exhaustive parameter requires all characters to be in unicode to pass, I do not believe it will work for shabados/desktop.

This is why I believe  and argue `toAscii` should not be "fixed". `toAscii` is doing its job, but due to the nature of ASCII, the replacements have to be done after and cannot be done before.

`isGurmukhi` should be used by implementations of `toAscii` and check before calling the `toAscii` function.

**Comment by bhajneet** [comment]
> This is why I believe and argue `toAscii` should not be "fixed". `toAscii` is doing its job, but due to the nature of ASCII, the replacements have to be done after and cannot be done before.
> 
> `isGurmukhi` should be used by implementations of `toAscii` and check before calling the `toAscii` function.

That would still fail my use case.

**Comment by Sarabveer** [comment]
> 
> 
> > This is why I believe and argue `toAscii` should not be "fixed". `toAscii` is doing its job, but due to the nature of ASCII, the replacements have to be done after and cannot be done before.
> > `isGurmukhi` should be used by implementations of `toAscii` and check before calling the `toAscii` function.
> 
> That would still fail my use case.

How so, every keystroke can be run through `isGurmukhi` without exhaustive. This issue would be out of scope for `gurmukhi-utils`.

**Comment by bhajneet** [comment]
How does this work for us if we're converting each keystroke input for search query?

image

**Comment by Sarabveer** [comment]
Why is this issue being raised now? The functionality you have shown is expected.

Input is supposed to be explicit, either Gurmukhi or ASCII. We agreed mixing would not be allowed.

File: test/isGurmukhi.spec.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    [ 'ਗੁਰਮੁਖੀ', true ],
+    [ 'ਮੈਂ ਗੁਰਮੁਖੀ ਵਿਚ ਲਿਖ ਰਿਹਾ ਹਾਂ।', true ],
+    [ 'ਲੜੀਵਾਰ​ਗੁਰਬਾਣੀ', true ], // Has U+200B, Zero Width Space
+    [ 'मैं हिंदी में लिख रहा हूँ।', false ],
+    [ 'میں شاہ رخ میں لکھ رہا ہوں۔', false ],
+    [ 'ਗੁਰਮੁਖੀ & English', true ],
+    [ 'English & ਗੁਰਮੁਖੀ', false ],
+  ]
+
+  lines.map( ( [ string, result ] ) => it( `should convert '${string}' to ${result}`, () => {
```
Comment (Harjot1Singh) [inline test/isGurmukhi.spec.js:16]: `it` message is wrong, this does not convert anything

**Pull request merged**